packages/template-retail-react-app/app/pages/checkout/index.jsx

```diff
@@ -5,37 +5,55 @@
  * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/BSD-3-Clause
  */
 import React, {useEffect, useState} from 'react'
-import {FormattedMessage} from 'react-intl'
+import {FormattedMessage, useIntl} from 'react-intl'
 import {Alert, AlertIcon, Box, Button, Container, Grid, GridItem, Stack} from '@chakra-ui/react'
 import useNavigation from '../../hooks/use-navigation'
 import {CheckoutProvider, useCheckout} from './util/checkout-context'
 import ContactInfo from './partials/contact-info'
 import ShippingAddress from './partials/shipping-address'
 import ShippingOptions from './partials/shipping-options'
-// import Payment from './partials/payment'
+import Payment from './partials/payment'
 import OrderSummary from '../../components/order-summary'
 import {useCurrentCustomer} from '../../hooks/use-current-customer'
 import {useCurrentBasket} from '../../hooks/use-current-basket'
 import CheckoutSkeleton from './partials/checkout-skeleton'
+import {useUsid, useShopperOrdersMutation} from 'commerce-sdk-react-preview'
 
 const Checkout = () => {
+    const {formatMessage} = useIntl()
     const navigate = useNavigation()
-    const {globalError, step, placeOrder} = useCheckout()
+    const usid = useUsid()
+    const {step} = useCheckout()
+    const [error, setError] = useState()
     const {data: basket} = useCurrentBasket()
     const [isLoading, setIsLoading] = useState(false)
-    // Scroll to the top when we get a global error
+    const {mutateAsync: createOrder} = useShopperOrdersMutation('createOrder')
+
     useEffect(() => {
-        if (globalError || step === 4) {
+        if (error || step === 4) {
             window.scrollTo({top: 0})
         }
-    }, [globalError, step])
+    }, [error, step])
 
     const submitOrder = async () => {
         setIsLoading(true)
         try {
-            await placeOrder()
-            navigate('/checkout/confirmation')
+            const order = await createOrder({
+                // We send the SLAS usid via this header. This is required by ECOM to map
+                // Einstein events sent via the API with the finishOrder event fired by ECOM
+                // when an Order transitions from Created to New status.
+                // Without this, various order conversion metrics will not appear on reports and dashboards
+                headers: {_sfdc_customer_id: usid},
+                body: {basketId: basket.basketId}
```

Comment:
```suggestion
                body: {basketId: basket?.basketId}
```
`basket` here can be undefined.

This is one of the downsides of calling `useCurrentBasket` to get the basket id. It does not feel as convenient as before. 

I believe previously things were less asynchronous. Now we're relying on react-query's built-in state/caching. But the tradeoff here is now we need to deal with the async stuff.

Reply:
this is a gotcha in general for React hooks. Not just for our stuff specifically... If our code is written in ts we would be in a lot more trouble... but for now, I'm not too worry about these. Especially those event handlers, if you can invoke them, basket must be loaded already, otherwise the loading state would stop users from invoking any actions, I think that's a safe assumption...

Reply:
Ideally our components should guard the undefined situation, but that's a lot of work to convert our codebase.

```js
const component = () => {
   const {data} = useBasket()
   if (!data) {
       return "loading..."
   }
   return ...
}
```

Reply:
> If our code is written in ts we would be in a lot more trouble
Well, what if the people using our hooks library start their project with TS?

I think it would be better if we save the current basket id inside a React Context instead.

Reply:
🤔 how is saving the basket id inside a React Context help in this case?

Reply:
Actually, my bad.. saving basket id inside react context still will not guarantee that it's defined.